Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift

```diff
@@ -441,6 +441,77 @@ class HTTPClientUpgradeTestCase: XCTestCase {
             .assertDoesNotContain(handlerType: NIOHTTPClientUpgradeHandler.self))
     }
     
+    func testUpgradeCompleteFlush() throws {
+        final class ChannelReadWriteHandler: ChannelDuplexHandler {
+            typealias OutboundIn = Any
+            typealias InboundIn = Any
+            typealias OutboundOut = Any
+            
+            var messagesReceived = 0
+            
+            func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+                self.messagesReceived += 1
+                context.writeAndFlush(data, promise: nil)
+            }
+        }
+
+        final class AddHandlerClientUpgrader<T: ChannelInboundHandler>: NIOHTTPClientProtocolUpgrader {
+            fileprivate let requiredUpgradeHeaders: [String] = []
+            fileprivate let supportedProtocol: String
+            fileprivate let handler: T
+
+            fileprivate init(forProtocol `protocol`: String, addingHandler handler: T) {
+                self.supportedProtocol = `protocol`
+                self.handler = handler
+            }
+            
+            func addCustom(upgradeRequestHeaders: inout HTTPHeaders) { }
+            
+            func shouldAllowUpgrade(upgradeResponse: HTTPResponseHead) -> Bool {
+                return true
+            }
+            
+            func upgrade(context: ChannelHandlerContext, upgradeResponse: HTTPResponseHead) -> EventLoopFuture<Void> {
+                return context.pipeline.addHandler(handler)
+            }
+        }
+        
+        var upgradeHandlerCallbackFired = false
+        let handler = ChannelReadWriteHandler()
+        let upgrader = AddHandlerClientUpgrader(forProtocol: "myproto", addingHandler: handler)
+        let clientChannel = try setUpClientChannel(clientHTTPHandler: ExplodingHTTPHandler(),
+                                                   clientUpgraders: [upgrader]) { (context) in
+                                                    
+                                                    upgradeHandlerCallbackFired = true
+        }
+        defer {
+            XCTAssertNoThrow(try clientChannel.finish())
+        }
+        
+        // Read the server request.
+        if let requestString = try clientChannel.readByteBufferOutputAsString() {
+                        
+            XCTAssertEqual(requestString, "GET / HTTP/1.1\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Length: 0\r\nConnection: upgrade\r\nUpgrade: myproto\r\n\r\n")
+        } else {
+            XCTFail()
+        }
```

Comment:
Under here add two more lines:

```
XCTAssertNoThrow(XCTAssertEqual(try clientChannel.readByteBufferOutputAsString(), ""))  // Empty body
XCTAssertNoThrow(XCTAssertNil(try clientChannel.readByteBufferOutputAsString()))
```

Then, please amend the code in `readByteBufferOutputAsString` in this file. Where we currently call `getString`, change it to `readString(length: requestBuffer.readableBytes)`, and change the binding of the buffer from `let` to `var`. Then things should start working.

Reply:
Awesome. Thanks!